### ISSUE TITLE
Add remote connection proxy parameters

### ISSUE THREAD

**Issue by mark0203** [SeleniumHQ/selenium#7247]
## 🚀 Feature Proposal

As a tester on a corporate network that is using a proxy to connect to the internet, I would like to have an option in selenium to add remote connection proxy parameters.

## Motivation

Our company is behind a proxy and therefore we have an issue when trying to connect to a remote selenium-grid that is outside of our network (e.g. browserstack/saucelabs) since the connection is getting refused (which seems to get stuck with urllib3).
For Java and Ruby I have found a workaround

Java
```
System.setProperty("https.proxyHost","192.168.1.10");
System.setProperty("https.proxyPort","3128");
```

Ruby
```
ENV['HTTP_PROXY'] = 'http://192.168.1.10:3128'
```

With Python I followed the instructions on browserstack:
https://www.browserstack.com/automate/python 
And I added this before starting the driver:
```
proxy = '192.168.1.10:3128'
os.environ['http_proxy'] = proxy 
os.environ['HTTP_PROXY'] = proxy
os.environ['https_proxy'] = proxy
os.environ['HTTPS_PROXY'] = proxy
```
But I keep getting stuck with `urllib3`
```
[ WARN ] Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x05FE20B0>: Failed to establish a new connection: [WinError 10060] A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond')': /wd/hub/session
```

## Example

To quote someone I spoke to:

> Looks indeed as if there isn't any more environment variable based injection of proxy settings possible for urllib3. You need to replace calls for urllib3.PoolManager(...) with urllib3.ProxyManager('http://localhost:8080',...)  or whatever your proxy_url is.
> Since Selenium doesn't has implemented an own setting for this within their API, you need to do this change on your own within site-packages\selenium\webdriver\remote\remote_connection.py

But that is going a bit to far for me and I rather see a solution like this:
```
driver = webdriver.Remote(
    command_executor='http://user:key@hub.browserstack.com:80/wd/hub',
    desired_capabilities=desired_cap,
    remote_http_proxy = '192.168.1.10:3128')
```
Or something you can add to the capabilities.
I propose a solution for Python since there is no easy workaround for that (or not that I understand). But it's not limited to Python. It would be great if we could use the same in Java and Ruby as well for example.

**Comment by jimevans** [comment]
I love how we want to use a given mechanism for Java, Ruby, and Python, but don’t say a thing about .NET or JavaScript...

(I’m really only joking here. Most open-source or Linux-or-MacOS-only folks would like to entirely forget about .NET, despite the relatively not-bad cross-platform story of .NET Core. No judgement; it was years before there was a decent cross-platform story, it’ll be years before people think of it in the same breath as Java or Python or Ruby for cross-platform work.)

**Comment by RajaAutomation** [comment]
@mark0203 Hi Mark, Can you please let me know of this feature is still being worked on or has it been implemented in any of the selenium releases.

**Comment by mark0203** [comment]
@jimevans I don't get the joke, feel free to elaborate on a more noob level since that's my level. Also feel free to mention workarounds for other program languages since this might help other people and maybe also contributes to indicating how much of a problem this really is. If it's only a `Python` and `urllib3` issue, it might have way less priority then when other languages also have issues like this.

@RajaAutomation I just submitted this request (2 days ago), it's a proposal, not a solution that can be found in one of the current releases if that is your question. If you ask if I'm working on it, the answer is "no", because I don't have the knowledge.

**Comment by jimevans** [comment]
@mark0203 My comment was in reaction to this part of your original post:

> But it's not limited to Python. It would be great if we could use the same in Java and Ruby as well for example.

That statement mentions two project-supplied language bindings, while omitting two others. And I’m particularly sensitive to the omission of .NET, as I’m the primary maintainer of those bindings. Many open-source advocates and users often forget about C# as a viable language for development because of the .NET Framework’s Windows-only origins, without recognizing the current state of .NET being open-source and cross-platform by its main vendor (Microsoft).

I apologize for the miscommunication. I will refrain from attempting humor in this forum going forward.

**Comment by mark0203** [comment]
@jimevans I also didn't mention: php, perl, javascript (did I forget any others? just assume I didn't mean to forget them), but that is not the point. The reason I put in this request is because it would be great if it could be resolved in selenium instead of in this external library that at least on python has no simple workaround. 
If you have a solution for any of the other languages, feel free to add those here. Than at least it will become clear if this is just a Python issue or if other languages have this issue as well with proxies.

**Comment by jimevans** [comment]
@mark0203 Of the languages you mentioned, only JavaScript (which I _did_ mention in my original attempt at humor) are supplied and maintained by the Selenium project. Again, I apologize for the miscommunication, and I will refrain from attempting humor in the future.

**Comment by mark0203** [comment]
It actually looks like this is the same issue I faced and it looks like they came up with the same conclusion (urllib3 is most likely the issue)
https://github.com/SeleniumHQ/selenium/issues/6264

**Comment by sharksforarms** [comment]
+1 for this, I had to change `PoolManager` to `ProxyManager` in `selenium/webdriver/remote/remote_connection.py`

https://urllib3.readthedocs.io/en/latest/advanced-usage.html#proxies

This is not ideal in an environment with a corporate proxy...

**Comment by sjieg** [comment]
https://github.com/seleniumhq/selenium/commit/8d98f3223c3757bac24f779f5d8c1e9e1e8edcfb
https://github.com/seleniumhq/selenium/commit/4534faa77c28674b6da8af26b0f6ed807017555e
@AutomatedTester is this something you could have a look at? Seems like these commits broke it.

**Comment by vanBadAsselt** [comment]
Yes, +1! . I also work in an environment with a corporate proxy. I had to downgrade to selenium 3.13.0 (where httplib is used instead of urllib3) in order to integrate Robot Framework tests with Browserstack Local. I would like to upgrade selenium again :)

If below is the solution (mentioned by @sharksforarms) then the proxy can maybe extracted from the system property or desiredcapabilities:

proxy = urllib3.ProxyManager('http://localhost:3128/', timeout=self._timeout)

**Comment by pekkaklarck** [comment]
As @mark0203 already noticed, this is a duplicate of #6264. The problem is that `urllib3` that Selenium nowadays uses doesn't support the semi standard `http_proxy` and `https_proxy` environment variables. Selenium apparently used something else (perhaps the standard `urllib`) that supported those environment variables until Selenium 3.14.

It seems to me that dropping the support for `http(s)_proxy` was not intentional but just a side effect of changing the underlying HTTP module. It also seems that nobody objects adding that support back, especially when other Selenium language bindings have such support. Someone should thus figure out how to do it (probably by checking `os.environ` and configuring `urllib3.ProxyManager` accordingly) and create a pull request.

**Comment by AutomatedTester** [comment]
Duplicate of #8053
